core_lib/src/managers/backupmanager.cpp

```diff
@@ -0,0 +1,523 @@
+                                                         mLayerId,
+                                                         editor());
+    mUndoStack->push(element);
+
+    emit updateBackup();
+}
+
+void BackupManager::layerRenamed()
+{
+    RenameLayerElement* element = new RenameLayerElement(mLayerName,
+                                                         mLayerId,
+                                                         editor());
+    mUndoStack->push(element);
+    emit updateBackup();
+}
+
+void BackupManager::cameraProperties(QRect backupViewRect)
+{
+    CameraPropertiesElement* element = new CameraPropertiesElement(mLayerName,
+                                                                   backupViewRect,
+                                                                   mLayerId,
+                                                                   editor());
+    mUndoStack->push(element);
+    emit updateBackup();
+}
+
+void BackupManager::frameDragged(int backupFrameOffset)
+{
+    DragFrameElement* element = new DragFrameElement(mLayerId,
+                                                     backupFrameOffset,
+                                                     editor());
+
+    mUndoStack->push(element);
+    emit updateBackup();
+}
+
+void BackupManager::flipView(bool backupIsFlipped, DIRECTION backupFlipDirection)
+{
+    FlipViewElement* element = new FlipViewElement(backupIsFlipped,
+                                                   backupFlipDirection,
+                                                   editor());
+
+    mUndoStack->push(element);
+    emit updateBackup();
+}
+
+void BackupManager::toggleSetting(bool /*backupToggleState*/, SETTING /*backupType*/)
+{
+//    ToggleSettingElement* element = new ToggleSettingElement(backupToggleState,
+//                                                             backupType,
+//                                                             editor());
+
+//    mUndoStack->push(element);
+//    emit updateBackup();
+}
+
+void BackupManager::layerMoved(int backupNewLayerIndex)
+{
+    MoveLayerElement* element = new MoveLayerElement(mLayerIndex,
+                                                     backupNewLayerIndex,
+                                                     editor());
+    mUndoStack->push(element);
+    emit updateBackup();
+}
+
+/**
+ * @brief BackupManager::prepareBackup
+ * This should be called prior to a modification
+ */
+void BackupManager::prepareBackup()
+{
+    mBitmap = nullptr;
+    mVector = nullptr;
+    mCamera = nullptr;
+    mClip = nullptr;
+    mKeyframe = nullptr;
+    mLayer = editor()->layers()->currentLayer();
+    mLayerId = mLayer->id();
+    mFrameIndex = editor()->currentFrame();
+    mIsSelected = editor()->getScribbleArea()->isSomethingSelected();
+    mSelectionRect = editor()->getScribbleArea()->mySelection;
+    mTempSelection = editor()->getScribbleArea()->myTempTransformedSelection;
```

Comment:
The coupling between BackupManager and ScribbleArea makes it impossible to write unit tests since there is no scribble area in the test program. However, let's make it works first and worry about unit tests later.

Reply:
Ah yeah you're probably right, we'll have to decouple it somehow... but yeah let's wait with unit testing till we have everything working